src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -4404,78 +4403,134 @@ private void GenerateOneCode()
                             Ldc(c);
                             MarkLabel(l4);
                         }
-                        Dup();
-                        Stloc(lenLocal);
-                        Ldc(1);
-                        Add();
-                        Stloc(cLocal);
 
-                        MarkLabel(l1);
-                        Ldloc(cLocal);
-                        Ldc(1);
-                        Sub();
-                        Dup();
-                        Stloc(cLocal);
-                        Ldc(0);
-                        if (Code() == RegexCode.Setloop || Code() == RegexCode.Setloopatomic)
-                        {
-                            BleFar(l2);
-                        }
-                        else
+                        // If this is a notoneloop{atomic} and we're left-to-right and case-sensitive,
+                        // we can use the vectorized IndexOf to search for the target character.
+                        if ((Code() == RegexCode.Notoneloop || Code() == RegexCode.Notoneloopatomic) &&
+                            !IsRightToLeft() &&
+                            !IsCaseInsensitive())
                         {
-                            Ble(l2);
-                        }
+                            Stloc(lenLocal);
 
-                        if (IsRightToLeft())
-                        {
-                            Leftcharnext();
-                        }
-                        else
-                        {
-                            Rightcharnext();
-                        }
+                            // i = runtext.AsSpan(runtextpos, len).IndexOf(ch);
+                            Ldloc(_runtextLocal!);
+                            Ldloc(_runtextposLocal!);
+                            Ldloc(lenLocal);
+                            Call(s_stringAsSpanIntIntMethod);
+                            Ldc(Operand(0));
+                            Call(s_spanIndexOf);
+                            Stloc(iLocal);
 
-                        if (Code() == RegexCode.Setloop || Code() == RegexCode.Setloopatomic)
-                        {
-                            EmitTimeoutCheck();
-                            EmitMatchCharacterClass(_strings![Operand(0)], IsCaseInsensitive(), charInClassLocal);
-                            BrtrueFar(l1);
+                            Label charFound = DefineLabel();
+
+                            // if (i != -1) goto charFound;
+                            Ldloc(iLocal);
+                            Ldc(-1);
+                            Bne(charFound);
```

Comment:
What's the most unlikely case here?  Could probably generate slightly shorter code if you outlined `Stloc(iLocal); BrFar(loopEnd)` by adding a short jump to those two instructions in the unlikely case.

Reply:
> What's the most unlikely case here?

It's hard to say; depends greatly on the input data.  My gut is that a) the majority case for this will be `.*`, b) it'll mostly be without the single line option such that `.` == `[^\n]`, and c) most inputs will have a newline, in which case this would find something.  But that's all speculation; probably not worth micro-optimizing for.